### PR TITLE
feature: add 'initialRecipeSelector' and SidebarRecipeSelector

### LINKED ISSUE #53: Dynamic UiRecipe selector
The goal: Each blueprint should only have one UiRecipe. That one recipe should define which sub-recipes should be selectable, and which component/plugin should be used to render the selection.

- [x] DMSS: Only return one UiRecipe
- [x] Create-dm-app should use the "UiPluginSelector" defined in the single recipe
- [x] dm-core: Create a simple alternative "UiPluginSelectorSideBar"

### PR DESCRIPTION
**Pull request by soofstad**
## What does this pull request change?
- Adds a new meta "RecipeSelector-Selector" - treat UiPluginSelector and UiPluginSideBarSelector as plugins
- Adds a new UiPluginSideBarSelector for concept demo and testing purposes

## Why is this pull request needed?
Need a generic "recipeSelector" that respects "initialRecipe"

## Issues related to this change
closes #53

### CODE REVIEW

File: packages/dm-core/src/components/UiRecipesSideBarSelector.tsx [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+type TSelectablePlugins = {
+  name: string
+  component: (props: IUIPlugin) => JSX.Element
+  config: any
+}
+
+export function UiRecipesSideBarSelector(props: IUIPlugin): JSX.Element {
+  const { idReference, categories, onSubmit, onOpen, type } = props
+  const { loading, getUiPlugin } = useContext(UiPluginContext)
+  const { uiRecipes, isLoading } = useBlueprint(type)
```
Comment (KristianKjerstad) [inline packages/dm-core/src/components/UiRecipesSideBarSelector.tsx:51]: suggestion: 
rename loading to isContextLoading
and isLoading to isBlueprintLoading

File: packages/dm-core/src/components/UiRecipesSideBarSelector.tsx [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+  useEffect(() => {
+    // Make sure ui plugins have been loaded
+    if (loading || isLoading) return
+    const recipes = uiRecipes.map((uiRecipe: any) => ({
+      name: uiRecipe?.label || uiRecipe?.name || uiRecipe?.plugin || 'No name',
+      component: getUiPlugin(uiRecipe?.plugin).component,
+      config: uiRecipe?.config,
+    }))
+    // @ts-ignore
```
Comment (KristianKjerstad) [inline packages/dm-core/src/components/UiRecipesSideBarSelector.tsx:65]: can this ts-ignore be deleted if we add a return type to the function inside the above map function?


```JavaScript
    const recipes = uiRecipes.map((uiRecipe: any): TSelectablePlugins => ({
      name: uiRecipe?.label || uiRecipe?.name || uiRecipe?.plugin || 'No name',
      component: getUiPlugin(uiRecipe?.plugin).component,
      config: uiRecipe?.config,
    }))
```

File: example/src/App.js [hunk 3]
```diff
@@ -38,22 +39,23 @@ function App() {
         }}
       >
         {loading ? (
-          <Loading />
+          <>Loading...</>
```
Comment (KristianKjerstad) [inline example/src/App.js:42]: Can replace with 
```JavaScript
      <div style={{ alignSelf: 'center', padding: '50px' }}>
        <CircularProgress color="primary" />
      </div>
```

but not a big deal

File: packages/dm-core/src/components/UiRecipesSelector.tsx [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+const Wrapper = styled.div`
+  align-self: start;
+  justify-content: space-evenly;
+  width: 100%;
+`
+
+export function UIRecipesSelector(props: IUIPlugin): JSX.Element {
+  const { idReference, type } = props
+  const { initialUiRecipe, isLoading, error } = useBlueprint(type)
+  const { loading, getUiPlugin } = useContext(UiPluginContext)
```
Comment (KristianKjerstad) [inline packages/dm-core/src/components/UiRecipesSelector.tsx:20]: suggestion:
rename loading to isContextLoading
and isLoading to isBlueprintLoading

File: packages/dm-core/src/components/UiRecipesSelector.tsx [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  const [initialPlugin, setInitialPlugin] = useState<any>()
+
+  useEffect(() => {
+    // Make sure uiRecipes has been loaded
+    if (isLoading || loading) return
+    let Component
```
Comment (KristianKjerstad) [inline packages/dm-core/src/components/UiRecipesSelector.tsx:26]:
```suggestion
    let component
```
camelcase

File: packages/dm-core/src/components/UiRecipesSelector.tsx [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+      case 'UiRecipesSideBarSelector':
+        Component = UiRecipesSideBarSelector
+        break
+      default:
+        Component = getUiPlugin(initialUiRecipe.plugin).component
+    }
+    setInitialPlugin({ component: Component })
+  }, [initialUiRecipe, isLoading, loading])
+
+  if (isLoading)
```
Comment (KristianKjerstad) [inline packages/dm-core/src/components/UiRecipesSelector.tsx:46]: should circular progress also be displayed when loading = True? (loading context)

File: packages/dm-core/src/components/UiPluginSelector.tsx [hunk 7]
```diff
@@ -149,24 +144,25 @@ function filterPlugins(
...
+    idReference,
     type,
     categories,
     breadcrumb,
     referencedBy,
     onSubmit,
     onOpen,
   } = props
-  const { uiRecipes, loadingBlueprint, error } = useBlueprint(type)
+  const { uiRecipes, isLoading, error } = useBlueprint(type)
```
Comment (KristianKjerstad) [inline packages/dm-core/src/components/UiPluginSelector.tsx:165]: suggestion:
rename loading to isContextLoading
and isLoading to isBlueprintLoading

**Review by KristianKjerstad — CHANGES_REQUESTED** [review]

File: example/src/App.js [hunk 8]
```diff
@@ -38,22 +39,23 @@ function App() {
...
         ) : (
-          <TreeView nodes={treeNodes} onSelect={() => {}} />
+          <TreeView
+            nodes={treeNodes}
+            onSelect={(node) => {
+              console.log(node)
```
Comment (eoaksnes) [inline example/src/App.js:47]: console.log

**Review by eoaksnes — APPROVED** [review]
Nice!

**Pull request merged**